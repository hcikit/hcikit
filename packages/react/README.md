# @hcikit/react

This package is a package that allows you to create and run HCI experiments. Visit hcikit.github.io for more information.
