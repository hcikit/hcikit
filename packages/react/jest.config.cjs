/** @type {import('@ts-jest/dist/types').InitialOptionsTsJest} */
module.exports = {
  preset: "ts-jest",
  testEnvironment: "jsdom",
  moduleNameMapper: {
    "@hcikit/workflow(.*)$": "<rootDir>/../workflow/src$1",
  },
  testMatch: ["./src/**/*.[jt]s?(x)", "**/?(*.)+(spec|test).[t]s?(x)"],
  modulePathIgnorePatterns: ["./dist/", "./node_modules/"],
  testPathIgnorePatterns: ["./dist/", "./node_modules/"],
  transformIgnorePatterns: ["./dist/", "/node_modules/"],
  setupFilesAfterEnv: ["<rootDir>/setupTests.ts"],
  collectCoverageFrom: [
    "src/**/*.{ts,tsx}",
    "!/node_modules/",
    "!src/setupTests.ts",
    "!src/**/*.stories.{js,ts,tsx,jsx}",
  ],
};
