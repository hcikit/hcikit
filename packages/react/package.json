{
  "name": "@hcikit/react",
  "version": "1.0.0-alpha.13",
  "description": "An experiment runner with a set of pre-built experiment components for running hci experiments",
  "author": "blainelewis1",
  "license": "MIT",
  "repository": "hcikit/hcikit",
  "module": "dist/index.js",
  "main": "dist/index.js",
  "type": "module",
  "keywords": [
    "react",
    "hci",
    "experiment"
  ],
  "types": "./dist/index.d.ts",
  "exports": {
    "import": "./dist/index.js",
    "default": "./dist/index.js"
  },
  "publishConfig": {
    "access": "public"
  },
  "files": [
    "dist",
    "README.md"
  ],
  "engines": {
    "node": ">=8",
    "npm": ">=5"
  },
  "scripts": {
    "test": "NODE_OPTIONS=--experimental-vm-modules yarn jest --env=jsdom",
    "test:watch": "NODE_OPTIONS=--experimental-vm-modules yarn jest --watch --env=jsdom --colors --verbose",
    "start": "tsc --watch",
    "build": "tsc --build --verbose",
    "prepack": "yarn run build",
    "storybook": "start-storybook -p 6006",
    "build-storybook": "build-storybook"
  },
  "peerDependencies": {
    "@mui/icons-material": "^5.5.0",
    "@mui/material": "^5.5.0",
    "@mui/styles": "^5.5.0",
    "react": "^18.2.0",
    "react-dom": "^18.2.0"
  },
  "dependencies": {
    "@hcikit/workflow": "workspace:*",
    "core-js": "^3.26.1",
    "lodash-es": "^4.17.21",
    "prop-types": "^15.8.1",
    "react-error-boundary": "^3.1.4",
    "react-is": "^18.2.0",
    "react-markdown": "^8.0.3",
    "serialize-error": "^11.0.0",
    "styled-components": "^5.3.6"
  },
  "gitHead": "8d0f2af41121f70996e8b5803d006832b147fd86",
  "devDependencies": {
    "@testing-library/jest-dom": "^5.16.5",
    "@testing-library/react": "^13.4.0",
    "@types/jest": "^29.2.3",
    "@types/react": "^18.0.25",
    "@types/styled-components": "^5.1.26",
    "@types/testing-library__jest-dom": "^5.14.5",
    "jest-diff": "^29.3.1",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "react-scripts": "^5.0.1",
    "typescript": "^4.9.3"
  }
}
