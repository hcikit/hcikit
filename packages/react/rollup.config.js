import babel from "rollup-plugin-babel";
import commonjs from "@rollup/plugin-commonjs";
import resolve from "@rollup/plugin-node-resolve";
import pkg from "./package.json";
import postcss from "rollup-plugin-postcss";

export default {
  input: "src/index.js",
  output: [
    {
      file: pkg.main,
      format: "cjs",
      sourcemap: true,
    },
    {
      file: pkg.module,
      format: "es",
      sourcemap: true,
    },
  ],
  external: Object.keys(pkg.peerDependencies || {}),
  plugins: [
    resolve(),
    commonjs({
      exclude: [
        "src/**",
        "../../node_modules/lodash-es/**",
        "../../node_modules/symbol-observable/es/**",
      ],
      include: ["../../node_modules/**"],
      namedExports: {
        "../../node_modules/react-redux/node_modules/react-is/index.js": [
          "isValidElementType",
          "isContextConsumer",
        ],
        "../../node_modules/@material-ui/utils/node_modules/react-is/index.js": [
          "ForwardRef",
          "Memo",
        ],
        "../../node_modules/react-dom/index.js": [
          "findDOMNode",
          "unstable_batchedUpdates",
        ],
        "../../node_modules/react-is/index.js": [
          "isValidElementType",
          "isElement",
          "ForwardRef",
          "typeOf",
          "Memo",
          "isContextConsumer",
        ],
        "../../node_modules/prop-types/index.js": ["element", "elementType"],
      },
    }),
    babel({
      exclude: "../../node_modules/**",
    }),
    postcss(),
  ],
};
