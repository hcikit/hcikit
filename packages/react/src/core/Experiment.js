import { throttle } from "lodash-es";
import React, { useEffect, useState } from "react";
import { Provider } from "react-redux";
import PropTypes from "prop-types";

import { configureStore } from "@hcikit/workflow";
import TaskRenderer from "./TaskRenderer";

const STATE_KEY = "HCIKIT_LOGS";

export const loadStateFromSessionStorage = () => {
  try {
    const state = window.sessionStorage.getItem(STATE_KEY);
    if (state) {
      return JSON.parse(state);
    }
  } catch (err) {
    console.error("Failed to load from sessionStorage", err);
  }

  return undefined;
};

export const saveStateToSessionStorage = throttle((state) => {
  try {
    window.sessionStorage.setItem(STATE_KEY, JSON.stringify(state));
  } catch (err) {
    console.error("Failed to save to sessionStorage", err);
  }
}, 3000);

let Experiment = ({
  saveState = saveStateToSessionStorage,
  loadState = loadStateFromSessionStorage,
  ...props
}) => {
  // TODO: maybe we should hide the task registry and instead just pass a list of objects?
  // TODO: not sure how to create different sessions for the same task.

  let [store, setStore] = useState();
  console.log(setStore);
  console.log({ store });
  /* eslint-disable react-hooks/exhaustive-deps */
  useEffect(() => {
    let storedState;

    if (process.env.NODE_ENV !== "development" && loadState) {
      storedState = loadState();
    }

    setStore(
      configureStore({ ...props.configuration, ...storedState }, saveState)
    );
  }, []);

  if (!store) {
    return null;
  }

  return (
    <Provider store={store}>
      <TaskRenderer {...props} />
    </Provider>
  );
};

Experiment.propTypes = {
  configuration: PropTypes.object.isRequired,
  loadState: PropTypes.func,
  saveState: PropTypes.func,
  tasks: PropTypes.objectOf(PropTypes.elementType),
};

export default Experiment;
