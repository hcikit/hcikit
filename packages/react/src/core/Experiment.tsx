import { isEmpty } from "lodash-es";
import React, {
  useState,
  createContext,
  useContext,
  useCallback,
  ElementType,
  useMemo,
  useEffect,
} from "react";

import { serializeError } from "serialize-error";
import TaskRenderer from "./TaskRenderer.js";

import {
  advanceConfiguration,
  logToConfiguration,
  modifyConfiguration,
  Configuration,
  ExperimentIndex,
  UnfilledLog,
  experimentComplete,
  getCurrentIndex,
  TypedConfiguration,
} from "@hcikit/workflow";

import GridLayout from "../GridLayout.js";

import { ErrorBoundary, FallbackProps } from "react-error-boundary";
import { CenteredNicePaper } from "../components/index.js";
import { BasePersistence, StoragePersistence } from "../persistence/index.js";

import { Button, Typography, CircularProgress } from "@mui/material";
import { Task } from "../index.js";

type PropsOfDict<T extends Record<string, React.ComponentType<any>>> = {
  [Task in keyof T]: React.ComponentProps<T[Task]>;
};

export type ConfigurationReact<
  T extends Record<string, React.ComponentType<any>>,
> = TypedConfiguration<PropsOfDict<T>>;

export interface ControlFunctions {
  advance: (index?: ExperimentIndex) => void;
  log: (log: UnfilledLog) => void;
  modify: (
    modifiedConfig: Record<string, unknown>,
    index?: ExperimentIndex
  ) => void;
  persistence?: BasePersistence | null;
}

export const ConfigContext = createContext<Configuration | undefined>(
  undefined
);

export const ConfigMutatorContext = createContext<ControlFunctions | undefined>(
  undefined
);

function useConfiguration(): Configuration {
  const context = useContext(ConfigContext);
  if (context === undefined) {
    throw new Error("useConfig must be used within a ConfigProvider");
  }
  return context;
}

function useExperiment(): ControlFunctions {
  const context = useContext(ConfigMutatorContext);

  if (context === undefined) {
    throw new Error("useConfig must be used within a ConfigMutatorContext");
  }

  return context;
}

export { useConfiguration, useExperiment };

function isPromise<T extends object>(obj: T | Promise<T>): obj is Promise<T> {
  return "then" in obj && typeof obj.then === "function";
}

const Experiment: React.FunctionComponent<{
  persistence?: BasePersistence | null;
  configuration: Configuration | Promise<Configuration>;
  tasks: Record<string, Task<any>>;
  Layout?: ElementType;
  ErrorHandler?: React.ComponentType<FallbackProps>;
  forceRemountEveryTask?: boolean;
  devOptions?: { persistInDevelopment: boolean; useIndexFromUrl: boolean };
}> = ({
  persistence = new StoragePersistence(window.sessionStorage, "HCIKIT_LOGS"),
  configuration: initialConfiguration,
  devOptions = { persistInDevelopment: false, useIndexFromUrl: true },
  ...props
}) => {
  const shouldPersistConfiguration =
    (process.env.NODE_ENV !== "development" ||
      devOptions.persistInDevelopment) &&
    persistence !== null;

  const [configurations, setConfigurations] = useState<{
    loadedConfiguration: Configuration | undefined;
    initialConfiguration: Configuration | undefined;
    doneLoadingPersisted: boolean;
    doneLoadingInitial: boolean;
  }>({
    loadedConfiguration: undefined,
    initialConfiguration: isPromise(initialConfiguration)
      ? undefined
      : initialConfiguration,
    doneLoadingPersisted: !shouldPersistConfiguration,
    doneLoadingInitial: !isPromise(initialConfiguration),
  });

  useEffect(() => {
    async function loadConfigs() {
      if (shouldPersistConfiguration) {
        persistence?.load().then((loadedConfig) => {
          setConfigurations((c) => ({
            ...c,
            loadedConfiguration: loadedConfig,
            doneLoadingPersisted: true,
          }));
        });
      }
      if (isPromise(initialConfiguration)) {
        initialConfiguration.then((loadedConfig) => {
          setConfigurations((c) => ({
            ...c,
            initialConfiguration: loadedConfig,
            doneLoadingInitial: true,
          }));
        });
      }
    }

    loadConfigs();
  }, []);

  if (
    !configurations.doneLoadingInitial ||
    !configurations.doneLoadingPersisted ||
    !configurations.initialConfiguration
  ) {
    return (
      <CenteredNicePaper>
        <Typography>Loading experiment configuration...</Typography>
        <br />
        <CircularProgress size={100} />
      </CenteredNicePaper>
    );
  } else {
    return (
      <ExperimentInner
        {...props}
        devOptions={devOptions}
        initialConfiguration={configurations.initialConfiguration}
        loadedConfiguration={configurations.loadedConfiguration}
        persistence={persistence}
      />
    );
  }
};

const ExperimentInner: React.FC<{
  persistence?: BasePersistence | null;
  initialConfiguration: Configuration;
  loadedConfiguration: Configuration | undefined;
  tasks: Record<string, Task>;
  Layout?: ElementType;
  ErrorHandler?: React.ComponentType<FallbackProps>;
  forceRemountEveryTask?: boolean;
  devOptions: { persistInDevelopment: boolean; useIndexFromUrl: boolean };
}> = ({
  persistence,
  tasks,
  Layout = GridLayout,
  ErrorHandler = DefaultErrorHandler,
  forceRemountEveryTask,
  initialConfiguration,
  loadedConfiguration,
  devOptions,
}) => {
  // TODO: not sure how to create different sessions for the same task. The issue is that they'll be overwritten by the other thing. Maybe we can add a config version or session key or something to it?
  // TODO: using config and configuration is so confusing...
  const [config, setConfig] = useState<Configuration>(() => {
    let configurationToUse: Configuration = initialConfiguration;

    if (
      loadedConfiguration &&
      (process.env.NODE_ENV !== "development" ||
        devOptions.persistInDevelopment)
    ) {
      configurationToUse = loadedConfiguration;
    }

    // This seems hacky. I think I am assigning the initial index to whatever it should be if the experiment just started, or assigning it to the proper one if we haven't started yet.
    configurationToUse = advanceConfiguration(
      configurationToUse,
      getCurrentIndex(configurationToUse)
    );

    return configurationToUse;
  });

  // it might be better to do this as we save state rather than waiting a render, I chose this way because it is easier to implement than adding it to each of the modifying functions below.
  // useEffect(() => {
  //   if (saveState) {
  //     saveState(config);
  //     console.log("saving");
  //   }
  // }, [config, saveState]);

  const advance = useCallback(
    (index?: ExperimentIndex) =>
      setConfig((c) => {
        c = logToConfiguration(c, { type: "END" });
        c = advanceConfiguration(c, index);

        if (!experimentComplete(c)) {
          c = logToConfiguration(c, { type: "START" });
        }

        persistence?.saveThrottled(c);

        return c;
      }),
    [persistence]
  );

  const log = useCallback(
    (log: UnfilledLog): void =>
      setConfig((c) => {
        const newConfig: Configuration = logToConfiguration(c, log);

        persistence?.saveThrottled(c);

        return newConfig;
      }),
    [persistence]
  );

  const modify = useCallback(
    (modifiedConfig: Record<string, unknown>, index?: ExperimentIndex): void =>
      setConfig((c) => {
        const newConfig = modifyConfiguration(c, modifiedConfig, index);

        persistence?.saveThrottled(c);

        return newConfig;
      }),
    [persistence]
  );

  // Kent says this is actually slower than just recreating it... But honestly idk if I trust that. Truth is I should test it, but this does solve the problem of unnecessary rerenders.

  const experiment = useMemo(
    () => ({
      advance: advance,
      log,
      modify,
      persistence,
    }),
    [advance, log, modify]
  );

  useEffect(() => {
    experiment.log({ type: "START" });
    // This is on purpose, we actually never want to rerun this. But also experiment should never actually change.
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useEffect(() => {
    window.onbeforeunload = (e) => {
      persistence?.save(config);
      persistence?.flush();
      return "You have some unsaved changes";
    };
  }, []);

  if (isEmpty(config)) {
    return null;
  }

  return (
    <ConfigMutatorContext.Provider value={experiment}>
      <ConfigContext.Provider value={config}>
        <ErrorBoundary
          onError={(error, info) =>
            log({
              type: "ERROR",
              ...serializeError(info),
              ...serializeError(error),
            })
          }
          FallbackComponent={ErrorHandler}
        >
          <Layout>
            {!experimentComplete(config) ? (
              <TaskRenderer
                tasks={tasks}
                forceRemountEveryTask={forceRemountEveryTask}
              />
            ) : (
              <div style={{ gridArea: "task" }}>
                <CenteredNicePaper>
                  <h2>You&apos;ve completed the experiment!</h2>
                  <a
                    download={`${config.participant || "log"}.json`}
                    href={`data:text/json;charset=utf-8,${encodeURIComponent(
                      JSON.stringify(config)
                    )}`}
                  >
                    Download experiment log
                  </a>
                </CenteredNicePaper>
              </div>
            )}
          </Layout>
        </ErrorBoundary>
      </ConfigContext.Provider>
    </ConfigMutatorContext.Provider>
  );
};

const DefaultErrorHandler: React.FunctionComponent<FallbackProps> = ({
  error,
}) => {
  const configuration = useConfiguration();
  return (
    <CenteredNicePaper>
      <div style={{ gridArea: "task" }}>
        <h2>An error occurred in the experiment.</h2>
        <p style={{ fontStyle: "italic" }}>
          ({error.name}: {error.message})
        </p>
        <br />

        <a
          download={`${configuration.participant || "log"}.json`}
          href={`data:text/json;charset=utf-8,${encodeURIComponent(
            JSON.stringify(configuration)
          )}`}
        >
          Download experiment log
        </a>
      </div>
    </CenteredNicePaper>
  );
};

export default Experiment;
