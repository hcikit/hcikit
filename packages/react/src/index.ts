import { PropsWithChildren, ReactElement, Component } from "react";
import { ControlFunctions } from "./core/Experiment.js";
import * as allTasks from "./tasks/index.js";

export * from "./tasks/index.js";
export * from "./components/index.js";

export { allTasks };

export { withGridItem } from "./GridLayout.js";

export { default as GridLayout } from "./GridLayout.js";
export { default } from "./core/Experiment.js";
export * from "./core/Experiment.js";
export {
  BasePersistence,
  StoragePersistence,
  FileSystemPersistence,
} from "./persistence/index.js";

type TaskProps = ControlFunctions;
type StaticProps = {
  layoutProps?: Record<string, any>;
};

export type Task<P = {}> = React.ComponentType<TaskProps & P> & StaticProps;

export interface FunctionTask<P = {}>
  extends React.FunctionComponent<P & ControlFunctions> {
  (
    props: PropsWithChildren<P & ControlFunctions>,
    context?: any
  ): ReactElement<any, any> | null;
}

export abstract class ClassTask<P = {}, S = {}, SS = any> extends Component<
  P & TaskProps,
  S,
  SS
> {}
