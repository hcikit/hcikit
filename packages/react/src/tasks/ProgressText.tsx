import React from "react";
import {
  getConfigurationAtIndex,
  indexToTaskNumber,
  getTotalTasks,
  getCurrentIndex,
  getLeafIndex,
} from "@hcikit/workflow";

import { withGridItem } from "../GridLayout.js";
import PropTypes from "prop-types";
import { useConfiguration } from "../core/Experiment.js";

// TODO: I updated the depth field to make a little bit more sense in the creativity project

const ProgressTextNoGrid: React.FunctionComponent<{ depth?: number }> = ({
  depth = 0,
}) => {
  let configuration = useConfiguration();

  const index = getCurrentIndex(configuration);
  const rightHalf = index.slice(depth);
  const leftHalf = index.slice(0, depth);

  configuration = getConfigurationAtIndex(configuration, leftHalf);
  const currentTaskNumber = indexToTaskNumber(
    configuration,
    getLeafIndex(configuration, rightHalf)
  );
  const totalTasks = getTotalTasks(configuration);

  return (
    <div>
      {currentTaskNumber + 1} / {totalTasks}
    </div>
  );
};

const ProgressText = withGridItem(ProgressTextNoGrid, "footer");

ProgressText.propTypes = {
  depth: PropTypes.number,
};

export { ProgressText };

export default ProgressText;
