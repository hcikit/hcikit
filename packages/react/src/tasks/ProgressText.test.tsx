import React from "react";
import ProgressText from "./ProgressText.js";
import { render, screen } from "@testing-library/react";
import Experiment, { useExperiment } from "../core/Experiment.js";
import { Configuration } from "@hcikit/workflow";
import userEvent from "@testing-library/user-event";

let BlankTask: React.FunctionComponent = () => null;

let basicConfig: Configuration = {
  __INDEX__: [0],
  tasks: ["ProgressText", "ClickToAdvance"],
  children: [
    { label: "hello world", task: "BlankTask" },
    { task: "BlankTask" },
    { label: "labels", task: "BlankTask" },
  ],
};

let advancedConfig: Configuration = {
  tasks: ["ProgressText", "ClickToAdvance"],
  depth: 1,

  children: [
    {
      label: "hello world",
      children: [
        { label: "0,0", task: "BlankTask" },
        { label: "0,1", task: "BlankTask" },
        { label: "0,2", task: "BlankTask" },
      ],
    },
    {
      label: "hello world",
      children: [
        { label: "1,0", task: "BlankTask" },
        { label: "1,1", task: "BlankTask" },
      ],
    },
    { label: "hello world", task: "BlankTask" },
  ],
};

let ClickToAdvance: React.FunctionComponent = () => {
  const experiment = useExperiment();
  return <button onClick={() => experiment.advance()}>advance</button>;
};

describe("ProgressText", () => {
  it("renders without crashing", () => {
    render(
      <Experiment
        tasks={{
          ClickToAdvance,
          ProgressText,
          BlankTask,
        }}
        configuration={{ ...basicConfig }}
        saveState={null}
        loadState={null}
      />
    );

    screen.getByText("1 / 3");
  });

  it("renders correctly selected task", async () => {
    render(
      <Experiment
        tasks={{
          ClickToAdvance,
          ProgressText,
          BlankTask,
        }}
        configuration={{ ...basicConfig }}
        saveState={null}
        loadState={null}
      />
    );

    screen.getByText("1 / 3");
    await userEvent.click(screen.getByText("advance"));
    screen.getByText("2 / 3");
    await userEvent.click(screen.getByText("advance"));
    screen.getByText("3 / 3");
  });

  it("renders correct depth parameter", async () => {
    render(
      <Experiment
        tasks={{
          ClickToAdvance,
          ProgressText,
          BlankTask,
        }}
        configuration={{ ...advancedConfig }}
        saveState={null}
        loadState={null}
      />
    );
    screen.getByText("1 / 3");
    await userEvent.click(screen.getByText("advance"));
    screen.getByText("2 / 3");
    await userEvent.click(screen.getByText("advance"));
    screen.getByText("3 / 3");
    await userEvent.click(screen.getByText("advance"));
    screen.getByText("1 / 2");
    await userEvent.click(screen.getByText("advance"));
    screen.getByText("2 / 2");
  });

  it("renders correctly", () => {
    expect(
      render(
        <Experiment
          tasks={{
            ClickToAdvance,
            ProgressText,
            BlankTask,
          }}
          configuration={{ ...advancedConfig }}
          saveState={null}
          loadState={null}
        />
      ).asFragment()
    ).toMatchSnapshot();
  });
});
