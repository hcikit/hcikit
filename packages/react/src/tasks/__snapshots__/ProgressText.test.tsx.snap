// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressText renders correctly 1`] = `
<DocumentFragment>
  <div
    id="grid-layout"
    style="display: grid; width: 100%; height: 100%; grid-template-columns: 1fr; grid-template-rows: min-content 1fr min-content; grid-template-areas: "header"
      "task"
      "footer";"
  >
    <div
      style="grid-area: footer;"
    >
      <div>
        1 / 3
      </div>
    </div>
    <button>
      advance
    </button>
  </div>
</DocumentFragment>
`;
